File: Data Structures & Algorithms/matrixDFS/submission-2.py
class Solution:
    def countPaths(self, grid: List[List[int]]) -> int:
        count = 0
        if not grid:
            return count

        rows = len(grid)
        cols = len(grid[0])

        def dfs(grid, r, c, visited):
            nonlocal count
            if min(r, c) < 0 or r == rows or c == cols or grid[r][c] == 1 or (r, c) in visited:
                return

            if r == rows - 1 and c == cols - 1:
                count += 1
                return

            visited.add((r, c))

            directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]
            for dr, dc in directions:
                dfs(grid, r + dr, c + dc, visited)
            
            visited.remove((r, c))

        dfs(grid, 0, 0, set())
        return count

File: Data Structures & Algorithms/island-perimeter/submission-2.ts
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    islandPerimeter(grid: number[][]): number {
        const ROWS = grid.length
        const COLS = grid[0].length
        const visited = new Set()

        function dfs(r: number, c: number): number {
            if (Math.min(r, c) < 0 || r >= ROWS || c >= COLS || grid[r][c] === 0) {
                return 1
            }

            if (visited.has(`#${r}#${c}`)) {
                return 0
            }

            visited.add(`#${r}#${c}`)

            return dfs(r + 1, c) + dfs(r, c + 1) + dfs(r - 1, c) + dfs(r, c - 1)

        }

        for (let r = 0; r < ROWS; r++){
            for (let c = 0; c < COLS; c++) {
                if (grid[r][c] === 1) {
                    return dfs(r, c)
                }
            }
        }
        return 0
    }
}
